core/src/main/java/com/datastax/oss/driver/api/core/metadata/SafeInitNodeStateListener.java

```diff
@@ -0,0 +1,188 @@
+ */
+package com.datastax.oss.driver.api.core.metadata;
+
+import com.datastax.oss.driver.api.core.session.Session;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.BiConsumer;
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * A node state listener wrapper that delays (or ignores) init events until after the session is
+ * ready.
+ *
+ * <p>By default, the driver calls node state events, such as {@link #onUp} and {@link #onAdd},
+ * before the session is ready; see {@link NodeStateListener#onSessionReady(Session)} for a detailed
+ * explanation. This can make things complicated if your listener implementation needs the session
+ * to process those events.
+ *
+ * <p>This class wraps another implementation to shield it from those details:
+ *
+ * <pre>
+ * NodeStateListener delegate = ... // your listener implementation
+ *
+ * SafeInitNodeStateListener wrapper =
+ *     new SafeInitNodeStateListener(delegate, true);
+ *
+ * CqlSession session = CqlSession.builder()
+ *     .withNodeStateListener(wrapper)
+ *     .build();
+ * </pre>
+ *
+ * With this setup, {@code delegate.onSessionReady} is guaranteed to be invoked first, before any
+ * other method. The second constructor argument indicates what to do with the method calls that
+ * were ignored before that:
+ *
+ * <ul>
+ *   <li>if {@code true}, they are recorded, and replayed to {@code delegate} immediately after
+ *       {@link #onSessionReady}. They are guaranteed to happen in the original order, and before
+ *       any post-initialization events.
+ *   <li>if {@code false}, they are discarded.
+ * </ul>
+ *
+ * @since 4.6.0
+ */
+public class SafeInitNodeStateListener implements NodeStateListener {
+
+  private final NodeStateListener delegate;
+  private final boolean replayInitEvents;
+
+  // Write lock: recording init events or setting sessionReady
+  // Read lock: reading init events or checking sessionReady
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+  @GuardedBy("lock")
+  private boolean sessionReady;
+
+  @GuardedBy("lock")
+  private final List<InitEvent> initEvents = new ArrayList<>();
+
+  /**
+   * Creates a new instance.
+   *
+   * @param delegate the wrapped listener, to which method invocations will be forwarded.
+   * @param replayInitEvents whether to record events during initialization and replay them to the
+   *     child listener once it's created, or just ignore them.
+   */
+  public SafeInitNodeStateListener(@NonNull NodeStateListener delegate, boolean replayInitEvents) {
+    this.delegate = Objects.requireNonNull(delegate);
+    this.replayInitEvents = replayInitEvents;
+  }
+
+  @Override
+  public void onSessionReady(@NonNull Session session) {
+    lock.writeLock().lock();
+    try {
+      sessionReady = true;
```

Comment:
Should we check `sessionReady` to prevent multiple invocations of the delegate's methods (especially double delivery of init events)? I know the driver won't do this, but application code might.

Reply:
Hrmm not sure if I follow, the boolean is flipped only once at the end of init, how could it prevent double delivery?

Reply:
I was just suggesting the following idiom:
```
lock.writeLock().lock();
try {
	if (!sessionReady) {
		sessionReady = true;
		delegate.onSessionReady(session);
		if (replayInitEvents) {
			for (InitEvent event : initEvents) {
				event.invoke(delegate);
			}
		}
	}
} finally {
	lock.writeLock().unlock();
}
```

Reply:
I see. Yeah I'll add it, as you noted it's not supposed to happen, but it doesn't hurt to have an extra check.